### PR TITLE
FIX: HLS fails on first EXT-X-KEY if not supported and doesn't check any others

### PR DESCRIPTION
**Pull request by matthuisman**
The below playlist will fail because the first EXT-X-KEY is SAMPLE-AES
Even though it has a valid key method (widevine) also.

This PR changes the behavior to instead look for VALID methods and fallback to invalid.
If a valid isn't found, it logs invalid and keeps trying.

It also makes a lot easier to add more supported methods.
Much easier checking for what you support, than checking for ever possible case you dont.

eg.
INFO: Unsupported encryption method: SAMPLE-AES
INFO: Unsupported encryption method: SAMPLE-AES
INFO: Supported encryption method found:SAMPLE-AES-CTR

Without the supported message, it appears it's failed.
IT also makes it easy to spot in a log when a stream is encrypted but no method is found
INFO: Unsupported encryption method: SAMPLE-AES
INFO: Unsupported encryption method: SAMPLE-AES
(no supported message)

It also means in the future, you just add in supported methods.
Not have to keep trying to add in non-supported methods.

There is also a commit to check KEYFORMAT before METHOD as the actual below playlist uses SAMPLE-AES instead of SAMPLE-AES-CTR for the widevine. KEYFORMAT is therefore a more reliable way to check.
```
#EXT-X-KEY:URI="skd://itunes.apple.com/p65435462/c1",KEYFORMAT="com.apple.streamingkeydelivery",KEYFORMATVERSIONS="1",METHOD=SAMPLE-AES
#EXT-X-KEY:URI="data:text/plain;charset=UTF-16;base64,vgEAAAEAAQC0ATwAVwBSAE0ASABFAEEARABFAFIAIAB4AG0AbABuAHMAPQAiAGgAdAB0AHAAOgAvAC8AcwBjAGgAZQBtAGEAcwAuAG0AaQBjAHIAbwBzAG8AZgB0AC4AYwBvAG0ALwBEAFIATQAvADIAMAAwADcALwAwADMALwBQAGwAYQB5AFIAZQBhAGQAeQBIAGUAYQBkAGUAcgAiACAAdgBlAHIAcwBpAG8AbgA9ACIANAAuADMALgAwAC4AMAAiAD4APABEAEEAVABBAD4APABQAFIATwBUAEUAQwBUAEkATgBGAE8APgA8AEsASQBEAFMAPgA8AEsASQBEACAAQQBMAEcASQBEAD0AIgBBAEUAUwBDAEIAQwAiACAAVgBBAEwAVQBFAD0AIgBBAEEAQQBBAEEATwBZAEQAUgBuAGQAagBNAFMAQQBnAEkAQwBBAGcASQBBAD0APQAiAD4APAAvAEsASQBEAD4APAAvAEsASQBEAFMAPgA8AC8AUABSAE8AVABFAEMAVABJAE4ARgBPAD4APAAvAEQAQQBUAEEAPgA8AC8AVwBSAE0ASABFAEEARABFAFIAPgA=",KEYFORMAT="com.microsoft.playready",KEYFORMATVERSIONS="1",METHOD=SAMPLE-AES
#EXT-X-KEY:URI="data:text/plain;base64,AAAAOHBzc2gAAAAA7e+LqXnWSs6jyCfc1R0h7QAAABgSEAAAAAAD5ndGYzEgICAgICBI88aJmwY=",KEYFORMAT="urn:uuid:edef8ba9-79d6-4ace-a3c8-27dcd51d21ed",KEYFORMATVERSIONS="1",METHOD=SAMPLE-AES-CTR
```

yes, i will need to squash all the commits :)

### CODE REVIEW

**Comment by matthuisman** [comment]
`#EXT-X-KEY:URI="data:text/plain;base64,AAAAOHBzc2gAAAAA7e+LqXnWSs6jyCfc1R0h7QAAABgSEAAAAAAD5ndGYzEgICAgICBI88aJmwY=",KEYFORMAT="urn:uuid:edef8ba9-79d6-4ace-a3c8-27dcd51d21ed",KEYFORMATVERSIONS="1",METHOD=SAMPLE-AES` should be parsed as Widevine

**Comment by matthuisman** [comment]
actually, this needs work

The parser fails as the first line has METHOD=SAMPLE-AES in it but no urn:uuid:edef8ba9-79d6-4ace-a3c8-27dcd51d21ed
The parser needs to check all lines before failing.

Even if the above was correct and had METHOD=SAMPLE-AES-CTR - it would still fail as the first line has SAMPLE-AES

**Comment by peak3d** [comment]
@matthuisman sorry, there has now much changed in hlskid, so you'll have to provide a new PR

File: src/parser/HLSTree.cpp [hunk 1]
```diff
@@ -137,12 +137,12 @@ int HLSTree::processEncryption(std::string baseUrl, std::map<std::string, std::s
 
     current_iv_ = m_decrypter->convertIV(map["IV"]);
 
-    Log(LOGLEVEL_INFO, "Supported encryption method found: AES-128");
+    Log(LOGLEVEL_DEBUG, "Supported encryption method found: AES-128");
     return ENCRYPTIONTYPE_AES128;
   }
 
   // NO VALID
-  Log(LOGLEVEL_WARNING, "Unsupported encryption method: %s with keyformat %s", map["METHOD"].c_str(), map["KEYFORMAT"].c_str());
```
Comment (peak3d) [inline src/parser/HLSTree.cpp]: should remain WARNING

File: src/parser/HLSTree.h [hunk 2]
```diff
@@ -32,11 +32,10 @@ namespace adaptive
   public:
     enum
     {
-      ENCRYPTIONTYPE_INVALID = 0,
```
Comment (peak3d) [inline src/parser/HLSTree.h]: cosmetic? It's INVALID for use in inputstream.adaptive

**Comment by matthuisman** [comment]
Should I base new PR on master?

**Comment by matthuisman** [comment]
Should I base new PR on master?

**Review by peak3d — CHANGES_REQUESTED** [review]
It's too hard for me too read, this PR should be splitted into 2, one for changes, and one for cosmetic.
Beside this 15 commits are not what we want to see in git history, pls. suqsh all your commits into one as long there are no reasons why to use separate commits

File: src/parser/HLSTree.h (hunk 2 above)
Comment (matthuisman) [reply src/parser/HLSTree.h]: i went for ENCRYPTIONTYPE_UNSUPPORTED (to cover UNKNOWN and invalid)
as with the new behavour, we are not actually checking for unspported.
Only checking for supported.
So any thing else would either be unknown or invalid - therefore UNSUPPORTED
It also matches the log message better

**Comment by matthuisman** [comment]
ill close this pull request and do a nicer one based on lastest commit in hlksid branch

**Pull request closed without merging**